Userland/Libraries/LibELF/Arch/aarch64/plt_trampoline.S

```diff
@@ -8,6 +9,56 @@
 .globl _plt_trampoline
 .hidden _plt_trampoline
 .type _plt_trampoline,@function
-_plt_trampoline: # (object, relocation_index)
-    # FIXME: Possibly incomplete.
-    ret
+
+// x16  = &got.plt[2]
+// x17  = &_plt_trampoline
+// [sp] = &got[relocation_index + 3]
+// [sp + 8] = return address
+_plt_trampoline:
+    mov x17, sp
+
+    // Save argument registers: x0-x7, v0-v7
+    stp x0, x1, [sp, -16]!
+    stp x2, x3, [sp, -16]!
+    stp x4, x5, [sp, -16]!
+    stp x6, x7, [sp, -16]!
+
+    stp q0, q1, [sp, -32]!
+    stp q2, q3, [sp, -32]!
+    stp q4, q5, [sp, -32]!
+    stp q6, q7, [sp, -32]!
+
+    // Load DynamicObject* from got.plt[1]
+    ldr x0, [x16, #-8]
+
+    // Calculate the rela.plt relocation offset
+    ldr x2, [x17]
+    sub x1, x2, x16
+    sub x1, x1, #8
+    // GOT entries are 8 bytes, but sizeof(Elf64_Rela) == 24, so multiply
+    // by 3 to get the relocation offset
+    lsl x2, x1, #1
+    add x1, x1, x2
+
+    bl _fixup_plt_entry
+
+    // Save the resolved function's address
+    mov x16, x0
+
+    // Restore argument registers
+    ldp q6, q7, [sp], #32
+    ldp q4, q5, [sp], #32
+    ldp q2, q3, [sp], #32
+    ldp q0, q1, [sp], #32
+
+    ldp x6, x7, [sp], #16
+    ldp x4, x5, [sp], #16
+    ldp x2, x3, [sp], #16
+    ldp x0, x1, [sp], #16
+
+    // Restore link register saved by the PLT stub
+    ldp xzr, x30, [sp], #16
```

Comment:
Why `ldp` with `xzr` instead of `ldr x30, [sp], #24`? Is 24 not encodable in an arm immediate?

Reply:
The value is stored at `sp + 8`, but I believe your suggestion would load from the top of the stack.

Reply:
Oh, you want to decrement so by 16, i see, makes sense.